### PR TITLE
ALSA: core: introduce INFO_NO_REWINDS - take 3

### PR DESCRIPTION
**Pull request by plbossart**
This is a tentative solution to address @tiwai's feedback that the previous solution only checked the case where the application used snd_pcm_rewind(). There was still a case where the application could modify the mmap'ed appl_ptr and request a 'SYNC_PTR' ioctl.

### CODE REVIEW

**Comment by plbossart** [comment]
update based on @tiwai's feedback that we should return an error code and unconditionally check the appl_ptr range.

**Comment by plbossart** [comment]
I'd need your reviews @kv2019i @ujfalusi @ranj063, or if you don't have time your ok to have me use your previous Reviewed-by tag from v2. I'd like this series to go in 5.16 and we are at rc5 already.

File: sound/core/pcm_lib.c [hunk 1]
```diff
@@ -2127,13 +2127,27 @@ int pcm_lib_apply_appl_ptr(struct snd_pcm_substream *substream,
...
 
 	if (appl_ptr >= runtime->boundary)
 		return -EINVAL;
+	/*
+	 * check if a rewind is requested by the application
+	 */
+	if (substream->runtime->info & SNDRV_PCM_INFO_NO_REWINDS) {
+		diff = appl_ptr - old_appl_ptr;
+		if (diff >= 0) {
+			if (diff > runtime->buffer_size)
```
Comment (ranj063) [inline sound/core/pcm_lib.c:2144]: I'm good with the patch @plbossart but just wondering how is it really possible with a rewind?

Comment (plbossart) [reply sound/core/pcm_lib.c:2144]: IDK, I copied the code from Takashi verbatim. this gives me a migraine. Maybe on capture?

Comment (ranj063) [reply sound/core/pcm_lib.c:2144]: rewind with capture? I cant even imagine that scenario.

Comment (plbossart) [reply sound/core/pcm_lib.c:2144]: agree, it's an application that would just drop the previous samples and redo processing. that would be silly indeed.

Comment (tiwai) [reply sound/core/pcm_lib.c:2144]: No, it's the case where applptr going back below 0 back to the boundary. Then apparently the diff becomes positive, but it's a rewind action.

**Pull request merged**